File: deploy/stacks/pipeline_stack.py
# importing modules
import os
from aws_cdk import (
    aws_ssm as ssm,
    pipelines,
    core as cdk,
    aws_codepipeline as codepipeline,
    aws_s3 as s3,
    aws_codepipeline_actions as codepipeline_actions,
    aws_iam as iam,
    aws_codebuild as codebuild,
    aws_sns as sns,
    aws_codestarnotifications as codestarnotifications
)
from .data_portal_status_page_stack import DataPortalStatusPageStack


class DataPortalStatusPageStage(cdk.Stage):
    def __init__(self, scope: cdk.Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        app_stage = self.node.try_get_context("app_stage")
        props = self.node.try_get_context("props")

        stack_name = props['app_stack_name']

        # Create stack defined on stacks folder
        DataPortalStatusPageStack(
            self,
            "StatusPage",
            stack_name=f"{stack_name}",
            tags={
                "stage": app_stage,
                "stack": f"pipeline-{stack_name}"
            }
        )


# Class for the CDK pipeline stack


class CdkPipelineStack(cdk.Stack):

    def __init__(self, scope: cdk.Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # Defining app stage
        app_stage = self.node.try_get_context("app_stage")
        props = self.node.try_get_context("props")

        # Load SSM parameter for GitHub repo (Created via Console)
        codestar_arn = ssm.StringParameter.from_string_parameter_attributes(self, "codestarArn",
                                                                            parameter_name="codestar_github_arn"
                                                                            ).string_value

        # Create S3 bucket for artifacts
        pipeline_artifact_bucket = s3.Bucket(
            self,
            "data-portal-status-page-artifact-bucket",
            bucket_name=props["pipeline_artifact_bucket_name"][app_stage],
            auto_delete_objects=True,
            removal_policy=cdk.RemovalPolicy.DESTROY,
            block_public_access=s3.BlockPublicAccess.BLOCK_ALL
        )

        # Create a pipeline for status page
        status_page_pipeline = codepipeline.Pipeline(
            self,
            "StatusPageCodePipeline",
            artifact_bucket=pipeline_artifact_bucket,
            restart_execution_on_update=True,
            cross_account_keys=False,
            pipeline_name=props["pipeline_name"][app_stage],
        )

        # Create codestar connection fileset
        code_pipeline_source = pipelines.CodePipelineSource.connection(
            repo_string=props["repository_source"],
            branch=props["branch_source"][app_stage],
            connection_arn=codestar_arn,
            trigger_on_push=True
        )

        # Grab code_pipeline_source artifact
        artifactMap = pipelines.ArtifactMap()
        source_artifact = artifactMap.to_code_pipeline(
            x=code_pipeline_source.primary_output
        )

        # Create A pipeline for cdk stack and react build
        self_mutate_pipeline = pipelines.CodePipeline(
            self,
            "CodePipeline",
            code_pipeline=status_page_pipeline,
            synth=pipelines.ShellStep(
                "CDKShellScript",
                input=code_pipeline_source,
                commands=[
                    "cdk synth",
                    "mkdir ./cfnnag_output",
                    "for template in $(find ./cdk.out -type f -maxdepth 2 -name '*.template.json'); do cp $template ./cfnnag_output; done",
                    "cfn_nag_scan --input-path ./cfnnag_output"
                ],
                install_commands=[
                    "cd deploy",
                    "npm install -g aws-cdk",
                    "gem install cfn-nag",
                    "pip install -r requirements.txt"
                ],
                primary_output_directory="deploy/cdk.out"
            )
        )

        # Deploy infrastructure
        self_mutate_pipeline.add_stage(
            DataPortalStatusPageStage(
                self,
                "DataPortalStatusPageStage",
            )
        )

        front_end_bucket_arn = s3.Bucket.from_bucket_name(
            self,
            "FrontEndBucket",
            bucket_name=props["client_bucket_name"][app_stage]
        ).bucket_arn

        self_mutate_pipeline.build_pipeline()

        # Create CodeBuild Project 
        react_codebuild_project = codebuild.PipelineProject(
            self,
            "ProjectReactBuild",
            check_secrets_in_plain_text_env_variables=False,
            description="The project from codebuild to build react project.",
            project_name="DataPortalStatusPageReactBuild",
            environment=codebuild.BuildEnvironment(
                build_image=codebuild.LinuxBuildImage.from_code_build_image_id("aws/codebuild/standard:7.0"),
            )
        )

        react_codebuild_project.add_to_role_policy(
            iam.PolicyStatement(
                actions=["ssm:GetParameter"],
                effect=iam.Effect.ALLOW,
                resources=[
                    "arn:aws:ssm:%s:%s:parameter/data_portal/status_page/*" % (
                        self.region, self.account),
                    "arn:aws:ssm:%s:%s:parameter/data_portal/client/*" % (
                        self.region, self.account)
                ]
            )
        )

        react_codebuild_project.add_to_role_policy(
            iam.PolicyStatement(
                actions=[
                    "s3:DeleteObject",
                    "s3:GetObject",
                    "s3:ListBucket",
                    "s3:PutObject"
                ],
                effect=iam.Effect.ALLOW,
                resources=[
                    front_end_bucket_arn,
                    front_end_bucket_arn + "/*"
                ]
            )
        )

        # CodeBuild for react build script
        react_build_actions = codepipeline_actions.CodeBuildAction(
            input=source_artifact,
            project=react_codebuild_project,
            check_secrets_in_plain_text_env_variables=False,
            environment_variables={
                "REACT_APP_REGION": codebuild.BuildEnvironmentVariable(
                    value="ap-southeast-2",
                    type=codebuild.BuildEnvironmentVariableType.PLAINTEXT
                ),
                "REACT_APP_BUCKET_NAME": codebuild.BuildEnvironmentVariable(
                    value=props["client_bucket_name"][app_stage],
                    type=codebuild.BuildEnvironmentVariableType.PLAINTEXT
                ),
                "REACT_APP_UMCCR_DOMAIN_NAME": codebuild.BuildEnvironmentVariable(
                    value="/hosted_zone/umccr/name",
                    type=codebuild.BuildEnvironmentVariableType.PARAMETER_STORE
                ),
                "REACT_APP_DATA_PORTAL_API_DOMAIN": codebuild.BuildEnvironmentVariable(
                    value="/data_portal/backend/api_domain_name",
                    type=codebuild.BuildEnvironmentVariableType.PARAMETER_STORE
                ),
                "REACT_APP_COG_USER_POOL_ID": codebuild.BuildEnvironmentVariable(
                    value="/data_portal/client/cog_user_pool_id",
                    type=codebuild.BuildEnvironmentVariableType.PARAMETER_STORE
                ),
                "REACT_APP_COG_APP_CLIENT_ID": codebuild.BuildEnvironmentVariable(
                    value="/data_portal/status_page/cog_app_client_id_stage",
                    type=codebuild.BuildEnvironmentVariableType.PARAMETER_STORE
                ),
                "REACT_APP_OAUTH_DOMAIN": codebuild.BuildEnvironmentVariable(
                    value="/data_portal/client/oauth_domain",
                    type=codebuild.BuildEnvironmentVariableType.PARAMETER_STORE
                ),
                "REACT_APP_OAUTH_REDIRECT_IN": codebuild.BuildEnvironmentVariable(
                    value="/data_portal/status_page/oauth_redirect_in_stage",
                    type=codebuild.BuildEnvironmentVariableType.PARAMETER_STORE
                ),
                "REACT_APP_OAUTH_REDIRECT_OUT": codebuild.BuildEnvironmentVariable(
                    value="/data_portal/status_page/oauth_redirect_out_stage",
                    type=codebuild.BuildEnvironmentVariableType.PARAMETER_STORE
                ),
            },
            type=codepipeline_actions.CodeBuildActionType.BUILD,
            action_name="ReactBuildAction"
        )

        # Create React Build app stage
        self_mutate_pipeline.pipeline.add_stage(
            stage_name="ReactBuild",
            actions=[
                react_build_actions
            ]
        )

        codebuild_build_invalidate_project = codebuild.PipelineProject(
            self,
            "CodebuildProjectInvalidateStatsPageCDNCache",
            project_name="InvalidateDataPortalStatusPageCDNCache",
            check_secrets_in_plain_text_env_variables=False,
            environment=codebuild.BuildEnvironment(
                build_image=codebuild.LinuxBuildImage.STANDARD_5_0
            ),
            build_spec=codebuild.BuildSpec.from_object({
                "version": "0.2",
                "phases": {
                    "build": {
                        "commands": [
                            # Find distribution ID from stack name
                            f"""DISTRIBUTION_ID=$(aws cloudformation describe-stacks """
                            f"""--stack-name {props['app_stack_name']} """
                            f"""--query 'Stacks[0].Outputs[?OutputKey==`CfnOutputCloudFrontDistributionId`].OutputValue' """
                            f"""--output text)""",

                            """aws cloudfront create-invalidation --distribution-id ${DISTRIBUTION_ID} --paths "/*" """
                        ]
                    }
                }
            }),
            timeout=cdk.Duration.minutes(5),
            queued_timeout=cdk.Duration.minutes(5)
        )

        # Add invalidate CDN role
        codebuild_build_invalidate_project.add_to_role_policy(
            iam.PolicyStatement(
                resources=[
                    f"arn:aws:cloudfront::{os.environ.get('CDK_DEFAULT_ACCOUNT')}:distribution/*"
                ],
                actions=["cloudfront:CreateInvalidation"]
            )
        )
        # Add describe stack role
        codebuild_build_invalidate_project.add_to_role_policy(
            iam.PolicyStatement(
                resources=[
                    f"arn:aws:cloudformation:ap-southeast-2:{os.environ.get('CDK_DEFAULT_ACCOUNT')}:"
                    f"stack/{props['app_stack_name']}/*"
                ],
                actions=["cloudformation:DescribeStacks"]
            )
        )

        # Reset Cache
        codebuild_action_clear_cache = codepipeline_actions.CodeBuildAction(
            action_name="InvalidateCloudFrontCache",
            project=codebuild_build_invalidate_project,
            check_secrets_in_plain_text_env_variables=False,
            input=source_artifact
        )

        # Invalidate CDN cache
        self_mutate_pipeline.pipeline.add_stage(
            stage_name="CleanUpStage",
            actions=[
                codebuild_action_clear_cache
            ]
        )

        # SSM parameter for AWS SNS ARN
        data_portal_notification_sns_arn = ssm.StringParameter.from_string_parameter_attributes(
            self,
            "DataPortalSNSArn",
            parameter_name="/data_portal/backend/notification_sns_topic_arn"
        ).string_value

        # SNS chatbot
        data_portal_sns_notification = sns.Topic.from_topic_arn(
            self,
            "DataPortalSNS",
            topic_arn=data_portal_notification_sns_arn
        )

        # Add Chatbot Notificaiton
        self_mutate_pipeline.pipeline.notify_on(
            "SlackNotificationStatusPage",
            target=data_portal_sns_notification,
            events=[
                codepipeline.PipelineNotificationEvents.PIPELINE_EXECUTION_FAILED,
                codepipeline.PipelineNotificationEvents.PIPELINE_EXECUTION_SUCCEEDED
            ],
            detail_type=codestarnotifications.DetailType.BASIC,
            enabled=True,
            notification_rule_name="SlackNotificationStatusPagePipeline"
        )

File: deploy-ts/lib/pipeline-stack.ts
import * as cdk from "aws-cdk-lib";
import * as ssm from "aws-cdk-lib/aws-ssm";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as codepipeline from "aws-cdk-lib/aws-codepipeline";
import * as codepipeline_actions from "aws-cdk-lib/aws-codepipeline-actions";
import * as iam from "aws-cdk-lib/aws-iam";
import * as codebuild from "aws-cdk-lib/aws-codebuild";
import * as sns from "aws-cdk-lib/aws-sns";
import * as codestarnotifications from "aws-cdk-lib/aws-codestarnotifications";
import * as pipelines from "aws-cdk-lib/pipelines";

import { Construct } from "constructs";
import { DataPortalStatusPageStack } from "./data-portal-status-page-stack";

interface DataPortalStatusPageStageProps extends cdk.StageProps {
  appStage: string;
  props: { [key: string]: any };
}

class DataPortalStatusPageStage extends cdk.Stage {
  constructor(
    scope: Construct,
    id: string,
    props: DataPortalStatusPageStageProps
  ) {
    super(scope, id, props);

    const { appStage, props: appProps } = props;
    const stackName = appProps["appStackName"];

    new DataPortalStatusPageStack(this, "StatusPage", {
      stackName: stackName,
      tags: {
        stage: appStage,
        stack: `pipeline-${stackName}`,
      },
      appStage: appStage,
      props: {
        clientBucketName: appProps["clientBucketName"],
        aliasDomainName: appProps["aliasDomainName"],
      },
    });
  }
}

interface CdkPipelineStackProps extends cdk.StackProps {
  appStage: string;
  props: { [key: string]: any };
}

export class CdkPipelineStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: CdkPipelineStackProps) {
    super(scope, id, props);

    const { appStage, props: appProps } = props;

    const codestarArn = ssm.StringParameter.valueForStringParameter(
      this,
      "codestar_github_arn"
    );

    const pipelineArtifactBucket = new s3.Bucket(
      this,
      "data-portal-status-page-artifact-bucket",
      {
        bucketName: appProps["pipelineArtifactBucketName"][appStage],
        autoDeleteObjects: true,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
        blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      }
    );

    const statusPagePipeline = new codepipeline.Pipeline(
      this,
      "StatusPageCodePipeline",
      {
        artifactBucket: pipelineArtifactBucket,
        restartExecutionOnUpdate: true,
        crossAccountKeys: false,
        pipelineName: appProps["pipelineName"][appStage],
      }
    );

    const codePipelineSource = pipelines.CodePipelineSource.connection(
      appProps["repository_source"],
      appProps["branchSource"][appStage],
      {
        connectionArn: codestarArn,
        triggerOnPush: true,
      }
    );

    const sourceArtifact = new codepipeline.Artifact();

    const selfMutatePipeline = new pipelines.CodePipeline(
      this,
      "CodePipeline",
      {
        codePipeline: statusPagePipeline,
        synth: new pipelines.ShellStep("CDKShellScript", {
          input: codePipelineSource,
          commands: [
            "cdk synth",
            "mkdir ./cfnnag_output",
            'for template in $(find ./cdk.out -type f -maxdepth 2 -name "*.template.json"); do cp $template ./cfnnag_output; done',
            "cfn_nag_scan --input-path ./cfnnag_output",
          ],
          installCommands: [
            "cd deploy",
            "npm install -g aws-cdk",
            "gem install cfn-nag",
            "npm install",
          ],
          primaryOutputDirectory: "deploy/cdk.out",
        }),
      }
    );

    selfMutatePipeline.addStage(
      new DataPortalStatusPageStage(this, "DataPortalStatusPageStage", {
        appStage: appStage,
        props: appProps,
      })
    );

    const frontEndBucketArn = s3.Bucket.fromBucketName(
      this,
      "FrontEndBucket",
      appProps["clientBucketName"][appStage]
    ).bucketArn;

    selfMutatePipeline.buildPipeline();

    const reactCodebuildProject = new codebuild.PipelineProject(
      this,
      "ProjectReactBuild",
      {
        description: "The project from codebuild to build react project.",
        projectName: "DataPortalStatusPageReactBuild",
        environment: {
          buildImage: codebuild.LinuxBuildImage.STANDARD_7_0,
        },
      }
    );

    reactCodebuildProject.addToRolePolicy(
      new iam.PolicyStatement({
        actions: ["ssm:GetParameter"],
        effect: iam.Effect.ALLOW,
        resources: [
          `arn:aws:ssm:${this.region}:${this.account}:parameter/data_portal/status_page/*`,
          `arn:aws:ssm:${this.region}:${this.account}:parameter/data_portal/client/*`,
        ],
      })
    );

    reactCodebuildProject.addToRolePolicy(
      new iam.PolicyStatement({
        actions: [
          "s3:DeleteObject",
          "s3:GetObject",
          "s3:ListBucket",
          "s3:PutObject",
        ],
        effect: iam.Effect.ALLOW,
        resources: [frontEndBucketArn, `${frontEndBucketArn}/*`],
      })
    );

    const reactBuildActions = new codepipeline_actions.CodeBuildAction({
      actionName: "ReactBuildAction",
      project: reactCodebuildProject,
      input: sourceArtifact,
      environmentVariables: {
        REACT_APP_REGION: {
          value: "ap-southeast-2",
          type: codebuild.BuildEnvironmentVariableType.PLAINTEXT,
        },
        REACT_APP_BUCKET_NAME: {
          value: appProps["clientBucketName"][appStage],
          type: codebuild.BuildEnvironmentVariableType.PLAINTEXT,
        },
        REACT_APP_UMCCR_DOMAIN_NAME: {
          value: "/hosted_zone/umccr/name",
          type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        },
        REACT_APP_DATA_PORTAL_API_DOMAIN: {
          value: "/data_portal/backend/api_domain_name",
          type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        },
        REACT_APP_COG_USER_POOL_ID: {
          value: "/data_portal/client/cog_user_pool_id",
          type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        },
        REACT_APP_COG_APP_CLIENT_ID: {
          value: "/data_portal/status_page/cog_app_client_id_stage",
          type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        },
        REACT_APP_OAUTH_DOMAIN: {
          value: "/data_portal/client/oauth_domain",
          type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        },
        REACT_APP_OAUTH_REDIRECT_IN: {
          value: "/data_portal/status_page/oauth_redirect_in_stage",
          type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        },
        REACT_APP_OAUTH_REDIRECT_OUT: {
          value: "/data_portal/status_page/oauth_redirect_out_stage",
          type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
        },
      },
      checkSecretsInPlainTextEnvVariables: false,
      type: codepipeline_actions.CodeBuildActionType.BUILD,
    });

    selfMutatePipeline.pipeline.addStage({
      stageName: "ReactBuild",
      actions: [reactBuildActions],
    });

    const codebuildBuildInvalidateProject = new codebuild.PipelineProject(
      this,
      "CodebuildProjectInvalidateStatsPageCDNCache",
      {
        projectName: "InvalidateDataPortalStatusPageCDNCache",
        environment: {
          buildImage: codebuild.LinuxBuildImage.STANDARD_5_0,
        },
        buildSpec: codebuild.BuildSpec.fromObject({
          version: "0.2",
          phases: {
            build: {
              commands: [
                `DISTRIBUTION_ID=$(aws cloudformation describe-stacks --stack-name ${appProps["appStackName"]} --query 'Stacks[0].Outputs[?OutputKey==\`CfnOutputCloudFrontDistributionId\`].OutputValue' --output text)`,
                'aws cloudfront create-invalidation --distribution-id ${DISTRIBUTION_ID} --paths "/*"',
              ],
            },
          },
        }),
        timeout: cdk.Duration.minutes(5),
        queuedTimeout: cdk.Duration.minutes(5),
      }
    );

    codebuildBuildInvalidateProject.addToRolePolicy(
      new iam.PolicyStatement({
        resources: [
          `arn:aws:cloudfront::${process.env.CDK_DEFAULT_ACCOUNT}:distribution/*`,
        ],
        actions: ["cloudfront:CreateInvalidation"],
      })
    );

    codebuildBuildInvalidateProject.addToRolePolicy(
      new iam.PolicyStatement({
        resources: [
          `arn:aws:cloudformation:ap-southeast-2:${process.env.CDK_DEFAULT_ACCOUNT}:stack/${appProps["appStackName"]}/*`,
        ],
        actions: ["cloudformation:DescribeStacks"],
      })
    );

    const codebuildActionClearCache = new codepipeline_actions.CodeBuildAction({
      actionName: "InvalidateCloudFrontCache",
      project: codebuildBuildInvalidateProject,
      input: sourceArtifact,
      checkSecretsInPlainTextEnvVariables: false,
    });

    selfMutatePipeline.pipeline.addStage({
      stageName: "CleanUpStage",
      actions: [codebuildActionClearCache],
    });

    const dataPortalNotificationSnsArn =
      ssm.StringParameter.valueForStringParameter(
        this,
        "/data_portal/backend/notification_sns_topic_arn"
      );

    const dataPortalSnsNotification = sns.Topic.fromTopicArn(
      this,
      "DataPortalSNS",
      dataPortalNotificationSnsArn
    );

    selfMutatePipeline.pipeline.notifyOn(
      "SlackNotificationStatusPage",
      dataPortalSnsNotification,
      {
        events: [
          codepipeline.PipelineNotificationEvents.PIPELINE_EXECUTION_FAILED,
          codepipeline.PipelineNotificationEvents.PIPELINE_EXECUTION_SUCCEEDED,
        ],
        detailType: codestarnotifications.DetailType.BASIC,
        enabled: true,
        notificationRuleName: "SlackNotificationStatusPagePipeline",
      }
    );
  }
}
